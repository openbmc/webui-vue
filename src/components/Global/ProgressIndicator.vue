<template>
  <ul class="progress-indicator">
    <li
      v-for="(step, index) in _steps"
      :key="index"
      :class="{
        complete: step.complete,
        'current-step': isCurrentStep(index),
      }"
    >
      <b-link :index="index">
        <status-icon v-if="step.complete" status="success" />
        <span v-else class="icon-indicator" aria-hidden="true"></span>
        <span v-if="step.complete" class="sr-only">
          {{ $t('global.status.complete') }}
        </span>
        <span v-if="isCurrentStep(index)" class="sr-only">
          {{ $t('global.status.currentStep') }}
        </span>
        <p>{{ step.label }}</p>
        <span class="progress-line"></span>
      </b-link>
    </li>
  </ul>
</template>

<script>
import StatusIcon from '@/components/Global/StatusIcon';

export default {
  name: 'ProgressIndicator',
  components: { StatusIcon },
  props: {
    start: {
      type: Number,
      default: 0,
    },
    steps: {
      type: Array,
      default: () => [],
      validator: (prop) => {
        return prop.every((step) =>
          Object.prototype.hasOwnProperty.call(step, 'label')
        );
      },
    },
  },
  data() {
    return {
      currentStep: this.start,
    };
  },
  computed: {
    _steps() {
      return this.steps.map((step, index) => {
        const { complete } = step;
        if (!complete && index >= this.start) {
          step.complete = false;
        } else {
          step.complete = true;
        }
        return step;
      });
    },
  },
  methods: {
    isCurrentStep(index) {
      return index === this.currentStep;
    },
    next() {
      if (this.currentStep === this.steps.length) return;
      this._steps[this.currentStep].complete = true;
      this.currentStep += 1;
    },
  },
};
</script>

<style lang="scss" scoped>
ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

li {
  position: relative;

  p {
    max-width: 10rem;
    vertical-align: top;
    display: inline-block;
  }

  a {
    text-decoration: none;
    color: theme-color('dark');
    cursor: default;
  }
}

.progress-line {
  top: 0;
  left: 0;
  height: 100%;
  width: 1px;
  background-color: $border-color;
  border: 1px inset transparent;
  position: absolute;
}

.current-step,
.complete {
  .progress-line {
    background-color: theme-color('primary');
  }
}

.status-icon {
  vertical-align: baseline;
  margin-left: 1rem;
  margin-right: 1rem;
}

.icon-indicator {
  border-color: theme-color('primary');
  border-radius: 50%;
  border-style: dashed;
  border-width: 2px;
  display: inline-block;
  height: 20px;
  margin-left: 1rem;
  margin-right: 1rem;
  position: relative;
  vertical-align: middle;
  width: 20px;
}

.current-step {
  .icon-indicator {
    border-style: solid;

    &::after {
      background-color: theme-color('primary');
      border-radius: 50%;
      content: '';
      display: inline-block;
      height: 14px;
      left: 1px;
      position: absolute;
      top: 1px;
      width: 14px;
    }
  }
}
</style>
