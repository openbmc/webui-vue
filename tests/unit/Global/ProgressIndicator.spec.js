import { mount, createLocalVue } from '@vue/test-utils';
import ProgressIndicator from '@/components/Global/ProgressIndicator';
import BootstrapVue from 'bootstrap-vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

describe('ProgressIndicator.vue', () => {
  const wrapper = mount(ProgressIndicator, {
    localVue,
    propsData: {
      steps: [
        { label: 'Step 1' },
        { label: 'Step 2' },
        { label: 'Step 3' },
        { label: 'Step 4' },
      ],
      start: 0,
    },
    mocks: {
      $t: (key) => key,
    },
  });

  wrapper.vm.next();
  wrapper.vm.next();
  wrapper.vm.showError();

  it('should exist', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should render 4 li elements', () => {
    expect(wrapper.findAll('li').length).toBe(4);
  });

  it('should render 2 completed steps', () => {
    expect(wrapper.findAll('.status-icon.success').length).toBe(2);
  });

  it('should render 1 error step', () => {
    expect(wrapper.findAll('.status-icon.danger').length).toBe(1);
  });

  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
