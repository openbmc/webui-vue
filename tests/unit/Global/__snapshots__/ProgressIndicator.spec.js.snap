// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressIndicator.vue should render correctly 1`] = `
<ul
  class="progress-indicator"
>
  <li
    class="complete"
  >
    <span
      class="status-icon success"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10,1c-4.9,0-9,4.1-9,9s4.1,9,9,9s9-4,9-9S15,1,10,1z M8.7,13.5l-3.2-3.2l1-1l2.2,2.2l4.8-4.8l1,1L8.7,13.5z"
        />
        <path
          d="M8.7,13.5l-3.2-3.2l1-1l2.2,2.2l4.8-4.8l1,1L8.7,13.5z"
          data-icon-path="inner-path"
          fill="none"
          opacity="0"
        />
      </svg>
    </span>
     
    <a
      class=""
      href="#"
      tabindex="-1"
      target="_self"
    >
      <span
        class="sr-only"
      >
        
        global.status.complete
      
      </span>
       
      <p>
        Step 1
      </p>
    </a>
     
    <span
      class="progress-line"
    />
  </li>
  <li
    class="complete"
  >
    <span
      class="status-icon success"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10,1c-4.9,0-9,4.1-9,9s4.1,9,9,9s9-4,9-9S15,1,10,1z M8.7,13.5l-3.2-3.2l1-1l2.2,2.2l4.8-4.8l1,1L8.7,13.5z"
        />
        <path
          d="M8.7,13.5l-3.2-3.2l1-1l2.2,2.2l4.8-4.8l1,1L8.7,13.5z"
          data-icon-path="inner-path"
          fill="none"
          opacity="0"
        />
      </svg>
    </span>
     
    <a
      class=""
      href="#"
      tabindex="-1"
      target="_self"
    >
      <span
        class="sr-only"
      >
        
        global.status.complete
      
      </span>
       
      <p>
        Step 2
      </p>
    </a>
     
    <span
      class="progress-line"
    />
  </li>
  <li
    class="current-step"
  >
    <span
      class="status-icon danger"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10,1c-4.9,0-9,4.1-9,9s4.1,9,9,9s9-4,9-9S15,1,10,1z M13.5,14.5L10,11l-3.5,3.5l-1-1L9,10L5.5,6.5l1-1L10,9l3.5-3.5l1,1	L11,10l3.5,3.5L13.5,14.5z"
        />
        <path
          d="M13.5,14.5L10,11l-3.5,3.5l-1-1L9,10L5.5,6.5l1-1L10,9l3.5-3.5l1,1L11,10l3.5,3.5L13.5,14.5z"
          data-icon-path="inner-path"
          fill="none"
          opacity="0"
        />
      </svg>
    </span>
     
    <a
      class=""
      href="#"
      tabindex="0"
      target="_self"
    >
      <span
        class="sr-only"
      >
        
        global.status.error
      
      </span>
       
      <p>
        Step 3
      </p>
    </a>
     
    <span
      class="progress-line"
    />
  </li>
  <li
    class=""
  >
    <span
      aria-hidden="true"
      class="icon-indicator"
    />
     
    <a
      class=""
      href="#"
      tabindex="-1"
      target="_self"
    >
      <!---->
       
      <p>
        Step 4
      </p>
    </a>
     
    <span
      class="progress-line"
    />
  </li>
</ul>
`;
